include/xtensor-blas/xlinalg.hpp

```diff
@@ -1514,6 +1514,172 @@ namespace linalg
         }
         return res;
     }
+
+    /**
+     * @brief Compute tensor dot product along specified axes for arrays
+     *
+     * Compute the sum of products along the last \em naxes axes of a and first
+     * \em naxes axes of b.
+     *
+     * @param xa input array
+     * @param xb input array
+     * @param naxes the number of axes to sum over
+     * @return resulting array
+     */
+    template <class T, class O>
+    auto tensordot(const xexpression<T>& xa, const xexpression<O>& xb, std::size_t naxes = 2)
+    {
+        using value_type = std::common_type_t<typename T::value_type, typename O::value_type>;
+        using result_type = std::conditional_t<T::static_layout == O::static_layout &&
+                                               (T::static_layout != layout_type::dynamic && T::static_layout != layout_type::any),
+                                               xarray<value_type, T::static_layout>,
+                                               xarray<value_type, XTENSOR_DEFAULT_LAYOUT>>;
+
+        result_type result;
+        const auto& a = xa.derived_cast();
+        const auto& b = xb.derived_cast();
+        if (naxes == 0)
+        {
+            // special case tensor outer product product
+            std::vector<std::size_t> result_shape(a.dimension() + b.dimension());
```

Comment:
it could be a good idea to use the `xt::dynamic_shape` typedef. The reason is that the `dynamic_shape`, by default, is a `svector`, which has a small buffer optimization and doesn't require allocation if the vector size stays < 4.

Reply:
Excellent, didn't know about `dyamic_shape` I fixed this.